Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -0,0 +1,165 @@
+        m_observer = observer.returnValue().ptr();
+    }
+    return m_observer.get();
+}
+
+void ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancyStatus>& relevancyToCheck)
+{
+    for (auto target : m_observer->observationTargets()) {
+        if (target) {
+            auto oldRelevancy = target->contentRelevancyStatus();
+            auto newRelevancy = oldRelevancy;
+            auto setRelevancyValue = [&](ContentRelevancyStatus reason, bool value) {
+                if (value)
+                    newRelevancy.add(reason);
+                else
+                    newRelevancy.remove(reason);
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::OnScreen))
+                setRelevancyValue(ContentRelevancyStatus::OnScreen, m_onScreenObservationTargets.contains(*target));
+
+            if (relevancyToCheck.contains(ContentRelevancyStatus::Focused))
+                setRelevancyValue(ContentRelevancyStatus::Focused, target->hasFocusWithin());
+
+            auto hasTopLayerinSubtree = [](const Element& target) {
+                for (auto& element : target.document().topLayerElements()) {
+                    if (element->isDescendantOf(target))
+                        return true;
+                }
+                return false;
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::IsInTopLayer))
+                setRelevancyValue(ContentRelevancyStatus::IsInTopLayer, hasTopLayerinSubtree(*target));
+
+            if (oldRelevancy == newRelevancy)
+                continue;
+            target->setContentRelevancyStatus(newRelevancy);
+            target->invalidateStyle();
+        }
+    }
+}
+
+// Workaround for lack of support for scroll anchoring. We make sure any content-visibility: auto elements
+// above the one to be scrolled to are already hidden, so the scroll position will not need to be adjusted
+// later.
+// FIXME: remove when scroll anchoring is implemented.
+void ContentVisibilityDocumentState::updateContentRelevancyStatusForScrollIfNeeded(const Element& scrollAnchor)
+{
+    if (!m_observer)
+        return;
+    auto findSkippedContentRoot = [](const Element& element) -> const Element* {
+        const Element* found = nullptr;
+        if (element.renderer() && element.renderer()->isSkippedContent()) {
+            for (auto candidate = &element; candidate; candidate = candidate->parentElementInComposedTree()) {
+                if (candidate->renderer() && candidate->renderStyle()->contentVisibility() == ContentVisibility::Auto)
+                    found = candidate;
+            }
+        }
+        return found;
+    };
+
+    if (auto *scrollAnchorRoot = findSkippedContentRoot(scrollAnchor)) {
+        for (auto target : m_observer->observationTargets()) {
+            if (target && target->renderer() && target->renderStyle()->contentVisibility() == ContentVisibility::Auto)
```

Comment:
Aren't all the observed renderers supposed to be content-visibility: auto ? Should this be an ASSERT?

Reply:
Good point, I replaced part of the if with an ASSERT.